### PR TITLE
[SYCL] Emit extra addrspace cast for @llvm.ptr.annotation

### PR DESCRIPTION
**Pull request by MrSidims**
When a field of a C++ struct used in a constructor, `this' resides in
    generic address space, and thus a pointer to the field also have
    generic address space.

    In this patch we ensure, that prior creating llvm.ptr.annotation call,
    we are aligning type of the intrinsic's argument with i8 private pointer.

    Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 1]
```diff
@@ -2136,15 +2136,12 @@ Address CodeGenFunction::EmitIntelFPGAFieldAnnotations(SourceLocation Location,
...
-      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, CGM.Int8PtrTy);
-  // FIXME Always emit the cast inst so we can differentiate between
-  // annotation on the first field of a struct and annotation on the struct
-  // itself.
-  if (VTy != CGM.Int8PtrTy)
-    V = Builder.CreateBitCast(V, CGM.Int8PtrTy);
+      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, VTy);
   V = EmitAnnotationCall(F, V, AnnotStr, Location);
-  V = Builder.CreateBitCast(V, VTy);
+  V = Builder.CreatePointerBitCastOrAddrSpaceCast(V, VTy);
```
Comment (asavonic) [inline clang/lib/CodeGen/CodeGenFunction.cpp]: Annotation intrinsic returns the same type that was passed in, so you don't need this line now.

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 2]
```diff
@@ -2136,15 +2136,12 @@ Address CodeGenFunction::EmitIntelFPGAFieldAnnotations(SourceLocation Location,
                                                        StringRef AnnotStr) {
   llvm::Value *V = Addr.getPointer();
   llvm::Type *VTy = V->getType();
+  assert(isa<llvm::PointerType>(VTy) &&
```
Comment (asavonic) [inline clang/lib/CodeGen/CodeGenFunction.cpp]: Can you use `cast` and make `VTy` a `PointerType` instead of this assert?

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 3]
```diff
@@ -16,9 +16,6 @@
 //CHECK: [[ANN8:@.str[\.]*[0-9]*]] = {{.*}}{memory:DEFAULT}{bankwidth:8}
 //CHECK: [[ANN9:@.str[\.]*[0-9]*]] = {{.*}}{memory:DEFAULT}{max_private_copies:4}
 
-//CHECK: @llvm.global.annotations
-//CHECK-SAME: a_one{{.*}}[[ANN1]]{{.*}}i32 159
```
Comment (asavonic) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Why did you remove this check?

**Comment by MrSidims** [comment]
> This patch uses a feature that is documented as illegal:
> 
> from LangRef:
> 
> > You can use llvm.ptr.annotation on a pointer to an _integer_ of any width.
> 
> This patch uses llvm.ptr.annotation on any types (float, user-defined structs, arrays?). Despite a correct declaration in Intrinsics.td, LLVM does not check for a pointee type, so this code does compile and pass Verifier, but I don't think that it is expected.
> 
> I suggest to at least follow up with the community regarding this use-case.

AFAIK the community is trying to move to type-less pointers and intrinsic, and that is being discussed in llvm-dev in "Typeless pointers and intrinsics" and "[RFC] A new multidimensional array indexing intrinsic" threads. But I agree with you, I will ask llvm-dev if it's a bug in verifier or an useful feature.

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 2 above)
Comment (MrSidims) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Well, I think I just don't need any of them.

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 1 above)
Comment (MrSidims) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Indeed, thanks!

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp (hunk 3 above)
Comment (MrSidims) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: By incident

**Comment by MrSidims** [comment]
Rebased to fix Windows build.

**Comment by bader** [comment]
@asavonic, are all your concerns addressed?

**Comment by asavonic** [comment]
> @asavonic, are all your concerns addressed?

No, the patch is still illegal, and we just get away with this.
Better way is to change the intrinsic in community (or add a new one) if we want to keep the original type. We'll have to do this anyway when typeless pointers in LLVM IR become a reality.

**Comment by asavonic** [comment]
In addition to that, if you want to proceed with this patch, you also need to make the same patch to SPIR-V Translator. Otherwise this change is not going to get through SPIR-V.

**Comment by MrSidims** [comment]
> > @asavonic, are all your concerns addressed?
> 
> No, the patch is still illegal, and we just get away with this.
> Better way is to change the intrinsic in community (or add a new one) if we want to keep the original type. We'll have to do this anyway when typeless pointers in LLVM IR become a reality.

The community doesn't react in "Typeless pointers and intrinsics" llvm-dev thread on my proposal.

> In addition to that, if you want to proceed with this patch, you also need to make the same patch to SPIR-V Translator. Otherwise this change is not going to get through SPIR-V.

It gets through SPIR-V.

**Comment by MrSidims** [comment]
@asavonic If I stay with unnecessary bit/addrspace casts surrounding the intrinsic will you be okay with that?

**Comment by asavonic** [comment]
> > In addition to that, if you want to proceed with this patch, you also need to make the same patch to SPIR-V Translator. Otherwise this change is not going to get through SPIR-V.
> 
> It gets through SPIR-V.

Usually these annotations are represented as type decorations in SPIR-V, not as instructions. If it gets through SPIR-V, then it means that SPIR-V translator emits similarly illegal instructions, right?

**Comment by MrSidims** [comment]
> > > In addition to that, if you want to proceed with this patch, you also need to make the same patch to SPIR-V Translator. Otherwise this change is not going to get through SPIR-V.
> > 
> > 
> > It gets through SPIR-V.
> 
> Usually these annotations are represented as type decorations in SPIR-V, not as instructions. If it gets through SPIR-V, then it means that SPIR-V translator emits similarly illegal instructions, right?

This instruction is being emitted with int8 pointer in private AS anyway disregarding the original type of the pointer... And actually I have no idea how to obtain it back. And so far there is no sense for the typeless intrinsic.

**Comment by asavonic** [comment]
> > > > In addition to that, if you want to proceed with this patch, you also need to make the same patch to SPIR-V Translator. Otherwise this change is not going to get through SPIR-V.
> > > 
> > > 
> > > It gets through SPIR-V.
> > 
> > 
> > Usually these annotations are represented as type decorations in SPIR-V, not as instructions. If it gets through SPIR-V, then it means that SPIR-V translator emits similarly illegal instructions, right?
> 
> This instruction is being emitted with int8 pointer in private AS anyway disregarding the original type of the pointer... And actually I have no idea how to obtain it back. And so far there is no sense for the typeless intrinsic.

Ok, so if the type we preserve here is going to vanish after SPIR-V, then it doesn't make much sense (unless you're willing to change SPIR-V translator as well).
An address space has to be fixed though.

**Comment by MrSidims** [comment]
Reworked. With this change a path in llvm-spirv translator is also needed.

**Comment by asavonic** [comment]
> Reworked. With this change a patch in llvm-spirv translator is also needed.

Can this patch be merged before the translator patch, or this has to wait until you fix the translator?

**Comment by MrSidims** [comment]
> > Reworked. With this change a patch in llvm-spirv translator is also needed.
> 
> Can this patch be merged before the translator patch, or this has to wait until you fix the translator?

Well, it will result in no regressions, since consumption from SPIR-V to LLVM IR is already broken for this case and only 2 patches together can fix it. So I would say we can do fixing incrementally.

**Comment by MrSidims** [comment]
Added patch in SPIR-V

File: llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll [hunk 4]
```diff
@@ -164,6 +169,71 @@ entry:
...
+
+for.cond.cleanup:                                 ; preds = %for.cond
+  %2 = bitcast i32* %i to i8*
+  call void @llvm.lifetime.end.p0i8(i64 4, i8* %2) #4
+  br label %for.end
+
+for.body:                                         ; preds = %for.cond
+  %3 = load i32, i32* %i, align 4, !tbaa !9
+  %mem = getelementptr inbounds %struct._ZTSZ20field_addrspace_castvE5state.state, %struct._ZTSZ20field_addrspace_castvE5state.state addrspace(4)* %this1, i32 0, i32 0
+  %4 = addrspacecast [8 x i32] addrspace(4)* %mem to i8*
```
Comment (asavonic) [inline llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: This seems very suspicious. Why do we cast from generic to private here?

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 5]
```diff
@@ -2142,14 +2142,10 @@ Address CodeGenFunction::EmitIntelFPGAFieldAnnotations(SourceLocation Location,
   llvm::Value *V = Addr.getPointer();
   llvm::Type *VTy = V->getType();
   llvm::Function *F =
-      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, CGM.Int8PtrTy);
-  // FIXME Always emit the cast inst so we can differentiate between
-  // annotation on the first field of a struct and annotation on the struct
-  // itself.
-  if (VTy != CGM.Int8PtrTy)
-    V = Builder.CreateBitCast(V, CGM.Int8PtrTy);
+      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, Int8PtrTy);
```
Comment (asavonic) [inline clang/lib/CodeGen/CodeGenFunction.cpp]: `llvm.ptr.annotation` supports pointers with address spaces, so it's better not to cast everything to `i8*`, but keep the address space by casting to `i8 addrspace(N)*`, where `N` matches with an address space of `V->getType()`.

File: llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll (hunk 4 above)
Comment (MrSidims) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: Done

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 5 above)
Comment (MrSidims) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Good point, thanks. Done.

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 6]
```diff
@@ -2141,15 +2141,13 @@ Address CodeGenFunction::EmitIntelFPGAFieldAnnotations(SourceLocation Location,
...
+  llvm::Type *Int8VPtrTy = llvm::Type::getInt8PtrTy(CGM.getLLVMContext(), AS);
   llvm::Function *F =
-      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, CGM.Int8PtrTy);
-  // FIXME Always emit the cast inst so we can differentiate between
-  // annotation on the first field of a struct and annotation on the struct
-  // itself.
-  if (VTy != CGM.Int8PtrTy)
-    V = Builder.CreateBitCast(V, CGM.Int8PtrTy);
+      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, Int8VPtrTy);
+  V = Builder.CreatePointerBitCastOrAddrSpaceCast(V, Int8VPtrTy);
```
Comment (asavonic) [inline clang/lib/CodeGen/CodeGenFunction.cpp]: If `V->getType()` and `Int8VPtrTy` have the same address space, then you don't need `CreatePointerBitCastOrAddrSpaceCast`, and `CreateBitCast` should suffice. You also don't need a bitcast if `V->getType()` is a pointer to an integer (of any width).

Comment (MrSidims) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Fixed

File: llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll (hunk 4 above)
Comment (asavonic) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: I'm confused. Do we still have the same generic to private cast again?

Comment (MrSidims) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: Ah, no, we don't have it anymore, I shall update the test.

File: llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll [hunk 7]
```diff
@@ -164,6 +169,71 @@ entry:
...
+  %5 = call i8* @llvm.ptr.annotation.p0i8(i8* %4, i8* getelementptr inbounds ([29 x i8], [29 x i8]* @.str.11, i32 0, i32 0), i8* getelementptr inbounds ([16 x i8], [16 x i8]* @.str.1, i32 0, i32 0), i32 24)
```
Comment (asavonic) [inline llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: Can you add a check that this instruction is emitted after SPIR-V -> LLVM IR translation.

Comment (MrSidims) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: I check it on lines 180-182.

I could add checks here as well, but for a c'tor the intrinsic is lost, which is considered to be a lesser bug to be fixed later.

Comment (asavonic) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: Ok, so you have 2 intrinsic calls before SPIR-V, and only 1 intrinsic call after SPIR-V. Can you add a FIXME note in place of a check?

Comment (MrSidims) [reply llvm-spirv/test/IntelFPGAMemoryAttributesForStruct.ll]: Sure. Done.

**Comment by MrSidims** [comment]
> Commit message seems to be a bit off. Did you update it for the latest revision of the patch?

I did. Now added some more information.

> Please split the patch to SPIR-V translator and commit it to upstream.

I have created PR to Khronos repo, but yesterday nightly there was broken. So I'd love to commit the patch here out of sync.

**Comment by bader** [comment]
@asavonic, all your comments are resolved. Please, dismiss "requested changes" review status to unblock the merge.
